drivers/auxdisplay/auxdisplay_tm1637.c

```diff
@@ -0,0 +1,380 @@
+
+	gpio_pin_set_dt(&cfg->data_pin, 1);
+	gpio_pin_set_dt(&cfg->clock_pin, 1);
+	tm1637_wait(dev);
+
+	gpio_pin_set_dt(&cfg->data_pin, 0);
+	tm1637_wait(dev);
+
+	gpio_pin_set_dt(&cfg->clock_pin, 0);
+	tm1637_wait(dev);
+}
+
+static void tm1637_stop_condition(const struct device *dev)
+{
+	const struct tm1637_config *cfg = dev->config;
+
+	gpio_pin_configure_dt(&cfg->data_pin, GPIO_OUTPUT);
+
+	gpio_pin_set_dt(&cfg->data_pin, 0);
+	gpio_pin_set_dt(&cfg->clock_pin, 1);
+	tm1637_wait(dev);
+
+	gpio_pin_set_dt(&cfg->data_pin, 1);
+	tm1637_wait(dev);
+}
+
+static bool tm1637_send_byte(const struct device *dev, uint8_t data_byte)
+{
+	const struct tm1637_config *cfg = dev->config;
+	bool ack;
+
+	for (int i = 0; i < 8; i++) {
+		gpio_pin_set_dt(&cfg->clock_pin, 0);
+		tm1637_wait(dev);
+
+		if (IS_BIT_SET(data_byte, i)) {
+			gpio_pin_configure_dt(&cfg->data_pin, GPIO_INPUT);
+		} else {
+			gpio_pin_configure_dt(&cfg->data_pin, GPIO_OUTPUT);
+			gpio_pin_set_dt(&cfg->data_pin, 0);
+		}
+		tm1637_wait(dev);
+
+		gpio_pin_set_dt(&cfg->clock_pin, 1);
+		tm1637_wait(dev);
+	}
+
+	/* Read ACK */
+	gpio_pin_set_dt(&cfg->clock_pin, 0);
+	gpio_pin_configure_dt(&cfg->data_pin, GPIO_INPUT);
+	tm1637_wait(dev);
+
+	gpio_pin_set_dt(&cfg->clock_pin, 1);
+	tm1637_wait(dev);
+
+	ack = !gpio_pin_get_dt(&cfg->data_pin);
+
+	if (!ack) {
+		gpio_pin_configure_dt(&cfg->data_pin, GPIO_OUTPUT);
+		gpio_pin_set_dt(&cfg->data_pin, 0);
+	}
+
+	tm1637_wait(dev);
+	gpio_pin_set_dt(&cfg->clock_pin, 0);
+
+	return ack;
+}
+
+static int tm1637_update_display(const struct device *dev)
+{
+	struct tm1637_data *data = dev->data;
+
+	/* Send data command */
+	tm1637_start_condition(dev);
+	tm1637_send_byte(dev, TM1637_CMD_DATA_AUTO_INC);
+	tm1637_stop_condition(dev);
+
+	/* Send segment data */
+	tm1637_start_condition(dev);
+	tm1637_send_byte(dev, TM1637_CMD_ADDR_BASE);
+
+	for (int i = 0; i < 4; i++) {
+		tm1637_send_byte(dev, data->display_buffer[i]);
+	}
+
+	tm1637_stop_condition(dev);
+
+	/* Send display control */
+	tm1637_start_condition(dev);
+	tm1637_send_byte(dev, TM1637_CMD_DISPLAY_CTRL | (data->current_brightness & 0x0F));
+	tm1637_stop_condition(dev);
+
+	return 0;
+}
+
+/* === auxdisplay driver API === */
+
+static int tm1637_auxdisplay_write(const struct device *dev, const uint8_t *buf, uint16_t len)
+{
+	struct tm1637_data *data = dev->data;
+	uint32_t pos = 0;
+	uint16_t i = 0;
+
+	/* Clear the display buffer first */
+	memset(data->display_buffer, 0, sizeof(data->display_buffer));
+
+	while (i < len && pos < 4) {
+		char c = buf[i];
+		uint8_t segment_code = 0;
+		bool valid_char = false;
+
+		if (c >= '0' && c <= '9') {
+			segment_code = segment_codes[c - '0'];
+			valid_char = true;
+		} else if (c >= 'A' && c <= 'Z') {
+			segment_code = segment_codes[c - 'A' + 10]; /* A starts at index 10 */
+			valid_char = true;
+		} else if (c >= 'a' && c <= 'z') {
+			segment_code = segment_codes[c - 'a' + 10]; /* a starts at index 10 */
+			valid_char = true;
```

Comment:
Using `+ 10` for letters is not very clear. It’d be better to separate digit and alphabet segment codes into different arrays for readability.